_firewall_config: add

rtollert opened the pull request:
Install firewalld and set up a default configuration which allows incoming ssh and wireguard and precious little else. Outgoing traffic permits http, https, wireguard, DHCP. wireguard interface is in "work" zone, everything else is in "public".

### Testing

Manually verified that both incoming and outgoing rules are being applied. Autotesting is currently TBD.

`nilrt-snac verify`, before configure:

```
Verifying firewall configuration...
(   57) ERROR nilrt_snac.verify: MISSING: running firewalld
(   57) ERROR nilrt_snac.verify: MISSING: firewall-cmd
(   57) ERROR nilrt_snac._check_target: ERROR: policy work-in target: expected REJECT, observed /bin/sh: line 1: firewall-cmd: command not found
(   58) ERROR nilrt_snac._check_target: ERROR: policy work-out target: expected REJECT, observed /bin/sh: line 1: firewall-cmd: command not found
(   59) ERROR nilrt_snac._check_target: ERROR: policy public-in target: expected REJECT, observed /bin/sh: line 1: firewall-cmd: command not found
(   59) ERROR nilrt_snac._check_target: ERROR: policy public-out target: expected REJECT, observed /bin/sh: line 1: firewall-cmd: command not found
```

`nilrt-snac configure`:
```
Configuring firewall...
```

`nilrt-snac verify`:
```
Verifying firewall configuration...
```


### Procedure

* [X] I certify that the contents of this pull request complies with the [Developer Certificate of Origin](https://github.com/ni/nilrt-snac/blob/master/docs/CONTRIBUTING.md).

rtollert commented:
@amstewart I don't know why the DCO check is failing. It's there? :F

AlexHearnNI commented on nilrt_snac/_configs/__init__.py:
```diff
@@ -30,4 +31,5 @@
     _TmuxConfig(),
     _PWQualityConfig(),
     _SudoConfig(),
+    _FirewallConfig()
```
```suggestion
    _FirewallConfig(),
```

rtollert replied:
fixed

AlexHearnNI commented on nilrt_snac/_configs/_firewall_config.py:
```diff
@@ -0,0 +1,101 @@
+from subprocess import run
+
+from nilrt_snac._configs._base_config import _BaseConfig
+from nilrt_snac._configs._config_file import _ConfigFile
```
```suggestion
```

AlexHearnNI commented on nilrt_snac/_configs/_firewall_config.py:
```diff
@@ -0,0 +1,101 @@
+                 "--set-target=REJECT"])
+
+            print("Firewall configuration complete (will take effect on reboot).")
+            # run(["firewall-cmd", "--reload"])
```
What is the purpose of this comment?

rtollert replied:
See below.

AlexHearnNI commented on nilrt_snac/_configs/_firewall_config.py:
```diff
@@ -0,0 +1,101 @@
+            # run(["firewall-cmd", "--reload"])
+
+    def verify(self, args: argparse.Namespace) -> bool:
+        return True
```
Is there nothing that can be done here? Is it not worth doing? Is it just not implemented yet? Please put a comment here explaining why this function is empty.

rtollert replied:
It's what would be needed to cause the firewall configuration to apply immediately. I wasn't sure if that's what we wanted, but it looks like quite a lot else applies immediately, and it works, so I'm just going to uncomment this and delete the print.

AlexHearnNI replied:
My comment was regarding the `verify` function.

rtollert commented:
v2:
- Addressed @AlexHearnNI's comments
- Added verify mode
- Added `check=True` for `subprocess.run()` calls
- Allow outbound dns

rtollert commented on nilrt_snac/_configs/_firewall_config.py:
```diff
@@ -0,0 +1,141 @@
+                        "--add-service=ssh",
+                        "--add-service=wireguard"],
+                       check=True)
+        subprocess.run(["firewall-offline-cmd",
```
I was thinking about this but only in the context of a `with` statement to prepend a list, at which point Python reminded me that's not a thing. I'll consider a helper function.

rtollert replied:
ok with refactoring this the next time around?

rtollert commented on nilrt_snac/_configs/_firewall_config.py:
```diff
@@ -0,0 +1,141 @@
+
+        valid: bool = True
+        valid = _check_target("work-in") and valid
+        valid = _check_target("work-out") and valid
```
To be honest I haven't fully thought out what verification should precisely mean; the checks here were mostly just stand-ins. I guess I got lucky.

Yes, there will be more thorough documentation for how exceptions etc. should work. In general, I would like to see firewalld's service facility used for this sort of thing whenever possible, with us supplying the service definitions.

AlexHearnNI commented on nilrt_snac/_configs/_firewall_config.py:
```diff
@@ -0,0 +1,141 @@
+        f"firewall-cmd --permanent --policy={policy} --get-target")
+    if expected == actual:
+        return True
+    logger.error(f"ERROR: policy {desc} target: expected {expected}, observed {actual}")
```
Is `desc` defined? What's the output when you run `verify` before `configure`?

rtollert replied:
Hm, it bails because firewall-cmd isn't installed, but yes, good catch. Will fix.

rtollert commented:
v3:
- Refactored firewall-offline-cmd calls re @dmondrik
- Fixed _check_target re @AlexHearnNI 
- Fixed execution before `configure` call re @AlexHearnNI
